numpy/core/src/multiarray/buffer.c

```diff
@@ -796,8 +796,10 @@ gentype_getbuffer(PyObject *self, Py_buffer *view, int flags)
     /* Fill in information */
     info = _buffer_get_info(self);
     if (info == NULL) {
-        PyErr_SetString(PyExc_BufferError,
+        if (!PyErr_Occurred()) {
```

Comment:
I don't think this `if` can ever be hit.

Reply:
Try setting a breakpoint there and calling `memoryview(np.datetime64('2016-01-01'))` or more to the point of the issue `pickle.dumps(np.datetime64('2016-01-01'))`. The call chain is `gettype_getbuffer` -> `_buffer_get_info` -> `_buffer_info_new` -> `_buffer_format_string` -> switch statement -> missing case marked with XXX -> raising error with "cannot include dtype '%c' in a buffer".

Reply:
To be clear - I think `_buffer_get_info` is well behaved, and will always set an exception if it returns `NULL`.

Reply:
ok